maven-surefire-common/src/main/java/org/apache/maven/surefire/providerapi/ProviderDetector.java

```diff
@@ -90,15 +95,13 @@
         }
     }
 
-    private ProviderInfo findByName( String providerClassName, ProviderInfo... wellKnownProviders )
+    @Nonnull
+    private ProviderInfo findByName( String providerClassName,
+                                     ConfigurableProviderInfo dynamicProvider, ProviderInfo... wellKnownProviders )
     {
-        for ( ProviderInfo wellKnownProvider : wellKnownProviders )
-        {
-            if ( wellKnownProvider.getProviderName().equals( providerClassName ) )
-            {
-                return wellKnownProvider;
-            }
-        }
-        return null;
+        return Arrays.stream( wellKnownProviders )
+            .filter( p -> p.getProviderName().equals( providerClassName ) )
+            .findFirst()
+            .orElseGet( () -> dynamicProvider.instantiate( providerClassName ) );
```

Comment:
`.orElseGet( () -> dynamicProvider.instantiate( providerClassName ) )` should be after the line 57.
This method is about finding something only.